frame/fnft/src/lib.rs

```diff
@@ -238,10 +243,60 @@ pub mod pallet {
 				Error::<T>::CollectionAlreadyExists
 			);
 			Collection::<T>::insert(collection, (who, admin, BTreeMap::<Vec<u8>, Vec<u8>>::new()));
+			Self::deposit_event(Event::FinancialNftCollectionCreated {
+				collection_id: *collection,
+				who: who.clone(),
+				admin: admin.clone(),
+			});
 			Ok(())
 		}
 	}
 
+	impl<T: Config> InspectEnumerable<T::AccountId> for Pallet<T> {
+		/// Returns an iterator of the collections in existence.
+		///
+		/// NOTE: iterating this list invokes a storage read per item.
+		fn collections() -> Box<dyn Iterator<Item = Self::CollectionId>> {
+			Box::new(Collection::<T>::iter_keys())
+		}
+
+		/// Returns an iterator of the items of a `collection` in existence.
+		///
+		/// NOTE: iterating this list invokes a storage read per item.
+		fn items(collection: &Self::CollectionId) -> Box<dyn Iterator<Item = Self::ItemId>> {
+			Box::new(
+				CollectionInstances::<T>::get(collection)
+					.into_iter()
+					.flat_map(|i| i.into_iter()),
+			)
+		}
+
+		/// Returns an iterator of the items of all collections owned by `who`.
+		///
+		/// NOTE: iterating this list invokes a storage read per item.
+		fn owned(
+			who: &T::AccountId,
+		) -> Box<dyn Iterator<Item = (Self::CollectionId, Self::ItemId)>> {
+			Box::new(OwnerInstances::<T>::get(who).into_iter().flat_map(|i| i.into_iter()))
+		}
+
+		/// Returns an iterator of the items of `collection` owned by `who`.
+		///
+		/// NOTE: iterating this list invokes a storage read per item.
+		fn owned_in_collection(
+			collection: &Self::CollectionId,
+			who: &T::AccountId,
+		) -> Box<dyn Iterator<Item = Self::ItemId>> {
+			let moved_collection = collection.clone();
+			Box::new(
+				OwnerInstances::<T>::get(who)
+					.into_iter()
+					.flat_map(|i| i.into_iter())
+					.filter_map(move |(c, i)| if c == moved_collection { Some(i) } else { None }),
```

Comment:
consider moving the `filter_map` before the `flat_map`,  slight performance optimization

Reply:
Not sure how that would be the case given that the number of elements filtered is the same?

Reply:
actually, disregard this. I misunderstood what the types were